src/Tools/dotnet-format/CodeFormatter.cs

```diff
@@ -0,0 +1,131 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.Formatting;
+using Roslyn.Utilities;
+using CSharpSyntaxKind = Microsoft.CodeAnalysis.CSharp.SyntaxKind;
+using VisualBasicSyntaxKind = Microsoft.CodeAnalysis.VisualBasic.SyntaxKind;
+using Microsoft.CodeAnalysis.Tools.CodeFormatter.MSBuild;
+using System.Diagnostics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.CodeAnalysis.Tools.CodeFormatter
+{
+    internal class CodeFormatter
+    {
+        public static async Task<int> FormatWorkspaceAsync(ILogger logger, string workspacePath, bool isSolution, CancellationToken token)
+        {
+            logger.LogInformation(string.Format(Resources.Info_FormattingFilesInWorkspace, Path.GetFileName(workspacePath)));
+
+            logger.LogTrace($"Loading workspace.");
+
+            var stopwatch = Stopwatch.StartNew();
+
+            var properties = new Dictionary<string, string>
+            {
+                // This property ensures that XAML files will be compiled in the current AppDomain
+                // rather than a separate one. Any tasks isolated in AppDomains or tasks that create
+                // AppDomains will likely not work due to https://github.com/Microsoft/MSBuildLocator/issues/16.
+                { "AlwaysCompileMarkupFilesInSeparateDomain", bool.FalseString },
+
+                // Use the latest language version to force the full set of available analyzers to run on the project.
+                { "LangVersion", "latest" },
+            };
+            using (var workspace = MSBuildWorkspace.Create(properties))
+            {
+                var projectPath = string.Empty;
+                if (isSolution)
+                {
+                    await workspace.OpenSolutionAsync(workspacePath, cancellationToken: token).ConfigureAwait(false);
+                }
+                else
+                {
+                    await workspace.OpenProjectAsync(workspacePath, cancellationToken: token).ConfigureAwait(false);
+                    projectPath = workspacePath;
+                }
+
+                logger.LogTrace($"Workspace loaded in {stopwatch.ElapsedMilliseconds}ms.");
+                stopwatch.Restart();
+
+                await FormatFilesInSolutionAsync(logger, workspace.CurrentSolution, projectPath, token).ConfigureAwait(false);
+
+                logger.LogTrace($"Format complete in {stopwatch.ElapsedMilliseconds}ms.");
+            }
+
+            logger.LogInformation(Resources.Info_FormatComplete);
+
+            return 0;
+        }
+
+        private static async Task FormatFilesInSolutionAsync(ILogger logger, Solution solution, string projectPath, CancellationToken cancellationToken)
+        {
+            foreach (var projectId in solution.ProjectIds)
```

Comment:
it seems like we could get a large speedup here by doing things in parallel.  even if we processed a single project at a time, we could format all docs in a single project at once, then we could apply all the results at the end of processing that project.

Reply:
I think really only syntax formatting can be done in parallel. Other actions, such as conforming to `_` prefix names, can't be done in parallel as edits can cross projects.

---
In reply to: [211515876](https://github.com/dotnet/roslyn/pull/29403#discussion_r211515876) [](ancestors = 211515876)

Reply:
Many semantic changes could not be done in parallel (due to dependencies).  However, naming is one of those that *can* be done in parallel as there aren't actually dependencies (between the edits that is) in that case.  i.e. if i have names X, Y and Z, i can rename those all to _X, _Y, _Z at the same time as any edits made from X->_X will be independent (i.e. will only touch a single token) from the other edits.

Reply:
Won't their be dependencies between renames because of shadowing issues? For instance a rename in Project A could require a rename in B to avoid shadowing at the same place B may be doing a rename? 

---
In reply to: [211676254](https://github.com/dotnet/roslyn/pull/29403#discussion_r211676254) [](ancestors = 211676254)

Reply:
@jaredpar I see.  yes, depending on your naming rules, there is the chance you could run into something there.

Though that does seem a bit unlikely...  But you're right that it cna't be totally ruled out :-/